### PR TITLE
Introduce mlflow.last_active_run fluent API

### PR DESCRIPTION
**Pull request by MarkYHZhang**
Signed-off-by: Mark Zhang <mark.zhang@databricks.com>

## What changes are proposed in this pull request?

Introduces a new fluent API `mlflow.last_active_run`. It returns `mlflow.active_run()` if it is not None. Otherwise, return the last run started from the current Python process that reached a terminal status (i.e. FINISHED, FAILED, or KILLED).

This is particularly useful for retrieving the most recently autologged run. Currently, there is no way of accessing the `mlflow.entities.Run` object of the most recent autolog run. This API resolves this inconvenience.

## How is this patch tested?

* Added tests in `tests/tracking/fluent/test_fluent.py` and `tests/tracking/fluent/test_fluent_autolog.py`
* Manually tested:
   * Retrieving the most recently autologged run without explicit `mlflow.start_run`
   * Retrieving the currently active run in between `mlflow.start_run` and `mlflow.end_run`
 

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

Introduces `mlflow.last_active_run()`: It returns the active run if one exists. Otherwise, it returns the last run started from the current Python process that reached a terminal status (i.e. FINISHED, FAILED, or KILLED). This is particularly useful for retrieving the most recent autolog run.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:395]: This should go in the :return: portion of the docstring

Comment (dbczumar) [reply mlflow/tracking/fluent.py:395]: For the first part of the docstring, I think we should simply say "Gets the most recent active run"

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:397]: Let's clarify that this is just one example where this is useful. Can we also include some example code?

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:407]: Did we identify the issue that prevented this from working in scikit-learn autologging tests? Also, have we audited all of our autologging integrations to make sure that they use fluent runs? Any autologging integration that uses `mlflow.autologging_utils.safe_patch(manage_run=True)` uses fluent runs, but others may not.

Comment (apurva-koti) [reply mlflow/tracking/fluent.py:407]: @MarkYHZhang and I looked through it and we reasoned that it wasn't an autologging issue, but merely the fact that in the following code
```
with mlflow.start_run() as r:
	mlflow.log_param("par", 3)
print(r.data.params)
> {}
```
`r` is populated at run creation time but the metrics/params aren't written back to that run object in memory after they've been sent off as API requests.
If this is unexpected behavior, then there's a big bug to fix!

Comment (MarkYHZhang) [reply mlflow/tracking/fluent.py:407]: Yeah the issue was because in that `test_fluent_autolog.py` file, there is an autouse enabled fixture that mocked out the post import hook.
https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/tests/tracking/fluent/test_fluent_autolog.py#L91-L94

Comment (dbczumar) [reply mlflow/tracking/fluent.py:407]: Cool!

File: mlflow/tracking/fluent.py
Comment (dbczumar) [reply mlflow/tracking/fluent.py:397]: Can we push changes before marking this resolved?

File: mlflow/tracking/fluent.py
Comment (dbczumar) [reply mlflow/tracking/fluent.py:395]: Can we push changes before marking this resolved?

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:435]:
```suggestion
    :return: The active run, if a run is currently active (this is equivalent to ``mlflow.active_run()``). Otherwise, there is no currently active run,
```

File: mlflow/tracking/fluent.py
Comment (dbczumar) [reply mlflow/tracking/fluent.py:444]: @MarkYHZhang Did we complete that audit?

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:435]:
```suggestion
    :return: The active run, if a run is currently active (this is equivalent to ``mlflow.active_run()``). Otherwise, if there is not currently an active run, the
```

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:392]: Can we also update our examples that use autologging to use this API, as well as add some docs about this here?: https://mlflow.org/docs/latest/tracking.html#automatic-logging

File: mlflow/tracking/fluent.py
Comment (MarkYHZhang) [reply mlflow/tracking/fluent.py:444]: Performing audit right now, it seems like all but the autologging in spark have `safe_patch` with `manage_run=True`.
https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/mlflow/spark.py#L842

Comment (MarkYHZhang) [reply mlflow/tracking/fluent.py:444]: But I ran this example, and it seems like `mlflow.last_active_run()` was able to return the correct run.
https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/mlflow/spark.py#L796-L824

Comment (dbczumar) [reply mlflow/tracking/fluent.py:444]: Fantastic! Spark autologging doesn't create runs, so that's all good. Can we introduce a test case in `test_fluent_autologging()` asserting that `safe_patch` is used with `manage_run=True` everywhere except Spark? We can probably add that to https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/tests/autologging/test_autologging_behaviors_integration.py#L86.

While we're at it, can we also add `mlflow.tensorflow` to https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/tests/autologging/test_autologging_behaviors_integration.py#L29?

File: mlflow/tracking/fluent.py
Comment (jinzhang21) [inline mlflow/tracking/fluent.py:415]: Is "auto_autolog" an official word? prefer autolog_run to keep it simple

**Review by jinzhang21 — APPROVED** [review]
LGTM! Please wait for @dbczumar 's LGTM before merging.

File: mlflow/tracking/fluent.py
Comment (dbczumar) [reply mlflow/tracking/fluent.py:415]: +1

File: tests/tracking/fluent/test_fluent_autolog.py
Comment (apurva-koti) [inline tests/tracking/fluent/test_fluent_autolog.py:352]: Ditto on making this `autologged_run`

File: tests/tracking/fluent/test_fluent.py
Comment (apurva-koti) [inline tests/tracking/fluent/test_fluent.py:1021]: Nittiest of nits, but could we have the expect match the input? ie make it `1.0` in line 1015, or `1` here

**Review by apurva-koti — APPROVED** [review]

File: docs/source/tracking.rst
Comment (dbczumar) [inline docs/source/tracking.rst:416]: Can we include a small example snippet?

File: tests/autologging/test_autologging_behaviors_integration.py
Comment (dbczumar) [inline tests/autologging/test_autologging_behaviors_integration.py:87]: Oof - thanks for removing this redundant for loop :)

File: tests/autologging/test_autologging_behaviors_integration.py
Comment (dbczumar) [inline tests/autologging/test_autologging_behaviors_integration.py:118]: Can we add an inline comment explaining why we're asserting this?

File: tests/tracking/fluent/test_fluent.py
Comment (dbczumar) [inline tests/tracking/fluent/test_fluent.py:1019]: Can we remove print statement?

File: tests/tracking/fluent/test_fluent_autolog.py
Comment (dbczumar) [inline tests/tracking/fluent/test_fluent_autolog.py:93]: Why is this necessary?

Comment (dbczumar) [reply tests/tracking/fluent/test_fluent_autolog.py:93]: Nevermind, I see.

**Review by dbczumar — APPROVED** [review]
LGTM with a couple tiny comments, thanks @MarkYHZhang !

File: tests/tracking/fluent/test_fluent_autolog.py
Comment (MarkYHZhang) [reply tests/tracking/fluent/test_fluent_autolog.py:93]: Recall that offline I mentioned how `mlflow.last_active_run` works in a notebook environment, but doesn't in test. The reason is because by mocking out `mlflow.tracking.fluent.register_post_import_hook`, the `_active_run_stack` in `fluent.py` does not get populated.

**Pull request merged**
